samples/lucy_bloom.cpp

```diff
@@ -0,0 +1,404 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "app/Config.h"
+#include "app/FilamentApp.h"
+#include "app/IBL.h"
+
+#include <filament/Engine.h>
+#include <filament/IndexBuffer.h>
+#include <filament/IndirectLight.h>
+#include <filament/LightManager.h>
+#include <filament/MaterialInstance.h>
+#include <filament/RenderableManager.h>
+#include <filament/RenderTarget.h>
+#include <filament/Scene.h>
+#include <filament/Skybox.h>
+#include <filament/TransformManager.h>
+#include <filament/VertexBuffer.h>
+#include <filament/View.h>
+
+#include <gltfio/AssetLoader.h>
+#include <gltfio/FilamentAsset.h>
+#include <gltfio/ResourceLoader.h>
+
+#include <getopt/getopt.h>
+
+#include <imgui.h>
+
+#include <string>
+
+#include "lucy_utils.h"
+
+using namespace filament;
+using namespace filament::math;
+using namespace gltfio;
+using namespace utils;
+
+static constexpr uint32_t FBO_WIDTH = 1024;
+static constexpr uint32_t FBO_HEIGHT = 1024;
+static const uint32_t WINDOW_WIDTH = 800;
+static const uint32_t WINDOW_HEIGHT = 800;
+
+struct Framebuffer {
+    Camera* camera = nullptr;
+    View* view = nullptr;
+    Scene* scene = nullptr;
+    Texture* color = nullptr;
+    Texture* depth = nullptr;
+    RenderTarget* target = nullptr;
+};
+
+struct LucyApp {
+    Config config;
+    bool showQuads;
+    bool showImgui;
+    float iblIntensity = 40000;
+    float iblRotation = M_PI * 1.5f;
+    AssetLoader* loader;
+    FilamentAsset* asset;
+    MaterialProvider* materials;
+    Entity rotationRoot;
+    Entity hudQuads[4];
+    Entity finalQuad;
+    Entity hblurQuad;
+    Entity vblurQuad;
+    Framebuffer reflection;
+    Framebuffer primary;
+    Framebuffer hblur;
+    Framebuffer vblur;
+    Camera* finalCamera;
+};
+
+static void setup(LucyApp& app, Engine* engine, View* finalView, Scene* finalScene) {
+
+    app.reflection.color = Texture::Builder()
+        .width(FBO_WIDTH).height(FBO_HEIGHT).levels(1)
+        .usage(Texture::Usage::COLOR_ATTACHMENT | Texture::Usage::SAMPLEABLE)
+        .format(Texture::InternalFormat::RGBA8).build(*engine);
+    app.reflection.depth = Texture::Builder()
+        .width(FBO_WIDTH).height(FBO_HEIGHT).levels(1)
+        .usage(Texture::Usage::DEPTH_ATTACHMENT)
+        .format(Texture::InternalFormat::DEPTH24).build(*engine);
+    app.reflection.target = RenderTarget::Builder()
+        .color(app.reflection.color) .depth(app.reflection.depth) .build(*engine);
+
+    app.primary.color = Texture::Builder()
+        .width(FBO_WIDTH).height(FBO_HEIGHT).levels(1)
+        .usage(Texture::Usage::COLOR_ATTACHMENT | Texture::Usage::SAMPLEABLE)
+        .format(Texture::InternalFormat::RGBA16F).build(*engine);
+    app.primary.depth = Texture::Builder()
+        .width(FBO_WIDTH).height(FBO_HEIGHT).levels(1)
+        .usage(Texture::Usage::DEPTH_ATTACHMENT)
```

Comment:
See `Texture::Builder::usage`:

```
Texture::Builder& Texture::Builder::usage(Texture::Usage usage) noexcept {
    // for now, the public API only allows UPLOADABLE and SAMPLEABLE textures
    mImpl->mUsage = Texture::Usage(Texture::Usage::DEFAULT | usage);
    return *this;
}
```

I think this should be updated before merging, otherwise the depth textures created here all have the UPLOADABLE usage bit.

Reply:
Excellent catch, fixed.